nupic/network/synapse.py

```diff
@@ -0,0 +1,38 @@
+﻿# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2014, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+
+
+class Synapse(object):

```

Comment:
In python and elsewhere there are similar abstractions, `list()` index access is generally faster than `dict()` key lookup, however, `list()` assumes a contiguous range whereas `dict()` can be random and sparse.  The trade-off is in memory.

You _can_ get better performance, as you suggest, by retaining and mutating an object, but you have to pre-allocate those objects AND cache them, which have a significant overhead.  Meanwhile in my example, in which the cache `dict` adds 400 Mb of memory overhead, if I needed to persist the object and reload it, that adds a substantial amount of overhead in writing to, and reading from disk.  The runtime benefits are a moot point if I have to load and save the cache object frequently, as we do in our own production settings.  The added memory footprint also limits concurrency.

It does seem redundant, however, to access `self._synapses[synapse]` and replace it with a new instance when it could possibly be changed in-place.  I'll comment in the issue with my thoughts.